### PR TITLE
Adding at() docs

### PR DESCRIPTION
**Pull request by Rumyra**
Adding array.at(), string.at() and typedarray.at() pages.

Looking at a review as these are actually my first js ref docs... some notes:

* Live samples won't work until this pr is merged https://github.com/mdn/interactive-examples/pull/1753/files (not sure why tests are failing)
* Note BCD data is not yet in
* Not sure what to reference in the way of spec which is here https://tc39.es/proposal-relative-indexing-method/

tl:dr; please review but don't merge until I've resolved above :) @chrisdavidmills

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+<h2 id="Syntax">Syntax</h2>
+
+<pre class="syntaxbox notranslate">var <var>item</var> = arr.at(index)</pre>
+
+<h3 id="Parameters">Parameters</h3>
+
+<dl>
+  <dt><code>index</code></dt>
+  <dd>The index (position) of the array element to be returned. Supports relative indexing from the end of the array when passed a negative index. i.e. If a negative number is used the element returned will be found by counting back from the end of the array.</dd>
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:34]:
```suggestion
  <dd>The index (position) of the array element to be returned. Supports relative indexing from the end of the array when passed a negative index; i.e. if a negative number is used, the element returned will be found by counting back from the end of the array.</dd>
```

File: files/en-us/web/javascript/reference/global_objects/string/at/index.html [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+<h2 id="Syntax">Syntax</h2>
+
+<pre class="syntaxbox notranslate">let <var>character</var> = string.at(index)</pre>
+
+<h3 id="Parameters">Parameters</h3>
+
+<dl>
+  <dt><code>index</code></dt>
+  <dd>The index (position) of the string character to be returned. Supports relative indexing from the end of the string when passed a negative index. i.e. If a negative number is used the character returned will be found by counting back from the end of the string.</dd>
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/string/at/index.html:29]:
```suggestion
  <dd>The index (position) of the string character to be returned. Supports relative indexing from the end of the string when passed a negative index; i.e. if a negative number is used, the character returned will be found by counting back from the end of the string.</dd>
```

**Review by sideshowbarker — APPROVED** [review]
Nits

**Review by sideshowbarker — APPROVED** [review]

**Comment by chrisdavidmills** [comment]
Hi @Rumyra !

I'm just preparing for the Fx 85 launch and giving this a review as part of it.

Looks mostly good. Any nitpicks I find I'll make directly to the text and not both you with them.

A few points to help you with the remainder:

* Looks like BCD still needs merging (just commented it). You might as well as the macro calls to thesepages now, so they are ready.
* You also need an entry in the Experimental features page, as this isn't in the release version yet.
* For the Spec, you need to use "ESNext — see https://github.com/mdn/yari/blob/master/kumascript/macros/SpecData.json#L777"

Can you make this a priority? The launch of Fx 85 is in 8 days. Thanks!

**Comment by Rumyra** [comment]
I've updated the spec links and bcd... one thing is the at method is not in the draft spec yet 😬 I've added what _should_ be the link...

**Comment by chrisdavidmills** [comment]
> one thing is the at method is not in the draft spec yet 😬 I've added what should be the link...

Hrm, this makes me wonder whether we shouldn't just link to the draft proposal instead for now.

@Elchi3 , what would you do here? `at()` is currently only specified at https://tc39.es/proposal-relative-indexing-method/, and not in the ESDraft spec yet, so we are not sure what to do about the spec link.

**Comment by Rumyra** [comment]
Let me know - I'll update here if it needs changing 👍

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+<ul>
+  <li>If you need to find an element based on a provided testing function, use {{jsxref("Array.prototype.find()", "find()")}}.</li>
+  <li>If you need to find if a value <strong>exists</strong> in an array, use {{jsxref("Array.prototype.includes()")}}.</li>
+  <li>If you need the <strong>index</strong> of an element in the array, use {{jsxref("Array.indexOf", "indexOf()")}}.</li>
+</ul>
+
+<h2 id="Syntax">Syntax</h2>
+
+<pre class="syntaxbox notranslate">var <var>item</var> = arr.at(index)</pre>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:28]: I'd remove the variable assignment in the syntax section. This section is mostly meant to show the method's signature. (also someone will probably come a long and say "Don't use var!".

```suggestion
<pre class="syntaxbox notranslate">arr.at(index)</pre>
```

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  - at
+  - Experimental
+---
+<div>{{JSRef}}</div>
+
+<p class="summary">The <strong><code>at()</code></strong> method takes an integer value and returns the item at that index, allowing for negative integers which count back from the last item.</p>
+
+<div>{{EmbedInteractiveExample("pages/js/array-at.html","shorter")}}</div>
+
+<ul>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:20]: Should this list have a heading or an opening text like "Usage notes:" or "Related methods"? Now it just appears after the interactive example without context.

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  <dt><code>index</code></dt>
+  <dd>The index (position) of the array element to be returned. Supports relative indexing from the end of the array when passed a negative index; i.e. if a negative number is used, the element returned will be found by counting back from the end of the array.</dd>
+</dl>
+
+<h3 id="Returns">Return value</h3>
+
+<p>The element in the array matching the given index. Returns {{jsxref('undefined')}} if the given index can not be found.</p>
+
+
+<h2 id="Polyfill">Polyfill</h2>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:42]: I'd not paste in a Polyfill. Maybe link to a polyfill in the "See also" section.

(for more of my reasoning, see https://discourse.mozilla.org/t/mdn-rfc-001-mdn-wiki-pages-shouldnt-be-a-distributor-of-polyfills/24500/14)

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  - Reference
+  - polyfill
+  - at
+  - Experimental
+---
+<div>{{JSRef}}</div>
+
+<p class="summary">The <strong><code>at()</code></strong> method takes an integer value and returns the item at that index, allowing for negative integers which count back from the last item.</p>
+
+<div>{{EmbedInteractiveExample("pages/js/array-at.html","shorter")}}</div>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:18]: Renders without scrollbar for me if you remove shorter.

```suggestion
<div>{{EmbedInteractiveExample("pages/js/array-at.html")}}</div>
```

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    Object.defineProperty(C.prototype, "at",
+                        { value: at,
+                          writable: true,
+                          enumerable: false,
+                          configurable: true });
+}</pre>
+
+<h2 id="Examples">Examples</h2>
+
+<h3>Return the last value of an array</h3>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:69]: I think an example showing this in comparison to slice/splice would be useful as people might want to see how `at()` reads compared to existing methods that achieve the same effect?

Also, in the spec repo it says that this finally gives "the ability to write `arr[-1]` instead of `arr[arr.length-1]`". So I think this page should mention somewhere that `at(-1)` is short for `arr[arr.length-1]` (which folks have used for years). What do you think?

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+<table class="standard-table">
+ <tbody>
+  <tr>
+   <th scope="col">Specification</th>
+   <th scope="col">Status</th>
+   <th scope="col">Comment</th>
+  </tr>
+  <tr>
+   <td>{{SpecName('ESDraft','#sec-array.prototype.at','Array.prototype.at')}}</td>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:102]: See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Pipeline_operator#specifications for how this is done for draft specs. Basically you add a new entry to [specdata](https://github.com/mdn/yari/blob/master/kumascript/macros/SpecData.json). I know this is painful and I wish you can just do `spec_url` in BCD and be done. That's my goal in https://github.com/mdn/content/issues/1146.

**Review by Elchi3 — CHANGES_REQUESTED** [review]
This is great! Happy to see you contributing to JS ref docs. I've reviewed this some more and left comments on the Array.prototype.at page for you, but they also apply to the other two methods mostly.

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 3 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:28]: Thanks @Elchi3 so just to clarify even if something is returned we don't tend to add var (or let or whatever stops peeps complaining 😆) - I've done this based on other examples.

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 4 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:20]: I took this from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find which was my inspiration for structure - I agree an h3 here would be good and I'm happy to update the find page too

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 5 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:42]: 👍

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 7 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:69]: Yes I do agree - I wondered this myself - I'll whip something up :)

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 8 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:102]: Oh thank you 🙏 I was wondering about this ☺️

**Comment by Rumyra** [comment]
Converting to draft until I've done feedbacks

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 4 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:20]: I've removed it as it's already in the see also section

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 8 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:102]: https://github.com/mdn/yari/pull/2637

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  - Reference
+  - polyfill
+  - at
+  - Experimental
+---
+<div>{{JSRef}}</div>
+
+<p class="summary">The <strong><code>at()</code></strong> method takes an integer value and returns the item at that index, allowing for positive and negative integers. Negative integers count back from the last item in the array.</p>
+
+<p>This is not to suggest there is anything wrong with using the square bracket notation. For example <code>array[0]</code> would return the first item. However instead of using {{domxref('Array.prototype.length','array.length')}} for latter items; e.g. <code>array[array.length-1]</code> for the last item, you can call <code>array.at(-1)</code>. <a href="#Examples">(See the examples below)</a></p>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:18]: Our macros are so weird. This could be a simple link, but no, instead you always have to remember our weird magic shortcuts. 😅 

```suggestion
<p>This is not to suggest there is anything wrong with using the square bracket notation. For example <code>array[0]</code> would return the first item. However instead of using {{jsxref('Array.prototype.length','array.length')}} for latter items; e.g. <code>array[array.length-1]</code> for the last item, you can call <code>array.at(-1)</code>. <a href="#Examples">(See the examples below)</a></p>
```

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+// Add an item to our 'cart' array
+cart.push('orange');
+const item2 = returnLast(cart);
+console.log(item2); // Logs: 'orange'
+</pre>
+
+<h3>Comparing methods</h3>
+
+<p>Here we compare different ways to select the penultimate (last but one) item of an {{domxref('Array')}}. Whilst all below methods are valid, it highlights the succinctness and readability of the <code>at()</code> method.</p>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/array/at/index.html:65]:
```suggestion
<p>Here we compare different ways to select the penultimate (last but one) item of an {{jsxref('Array')}}. Whilst all below methods are valid, it highlights the succinctness and readability of the <code>at()</code> method.</p>
```

File: files/en-us/web/javascript/reference/global_objects/string/at/index.html [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  - Reference
+  - polyfill
+  - at
+  - Experimental
+---
+<div>{{JSRef}}</div>
+
+<p class="summary">The <strong><code>at()</code></strong> method takes an integer value and returns a new {{jsxref('String')}} consisting of the single UTF-16 code unit located at the specified offset. This method allows for positive and negative integers. Negative integers count back from the last string character.</p>
+
+<div>{{EmbedInteractiveExample("pages/js/string-at.html","shorter")}}</div>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/string/at/index.html:18]:
```suggestion
<div>{{EmbedInteractiveExample("pages/js/string-at.html")}}</div>
```

File: files/en-us/web/javascript/reference/global_objects/string/at/index.html [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+invoiceRef = 'myinvoice02';
+
+console.log( returnLast(invoiceRef) );
+// Logs: '2'
+</pre>
+
+<h3>Comparing methods</h3>
+
+<p>Here we compare different ways to select the penultimate (last but one) character of a {{domxref('String')}}. Whilst all below methods are valid, it highlights the succinctness and readability of the <code>at()</code> method.</p>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/string/at/index.html:61]:
```suggestion
<p>Here we compare different ways to select the penultimate (last but one) character of a {{jsxref('String')}}. Whilst all below methods are valid, it highlights the succinctness and readability of the <code>at()</code> method.</p>
```

File: files/en-us/web/javascript/reference/global_objects/string/at/index.html [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+<pre class="brush: js notranslate">
+const myString = 'Every green bus drives fast.';
+
+// Using length property and charAt() method
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/string/at/index.html:66]: I guess the `charAt` comparison is the important one here, nice. We might want to point to at() from the charAt() page but can be a follow-up. Not mission-critical for this PR.

File: files/en-us/web/javascript/reference/global_objects/typedarray/at/index.html [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+  - Reference
+  - polyfill
+  - at
+  - Experimental
+---
+<div>{{JSRef}}</div>
+
+<p class="summary">The <strong><code>at()</code></strong> method takes an integer value and returns the item at that index, allowing for positive and negative integers. Negative integers count back from the last item in the array.</p>
+
+<p>This is not to suggest there is anything wrong with using the square bracket notation. For example <code>array[0]</code> would return the first item. However instead of using {{domxref('TypedArray.prototype.length','array.length')}} for latter items; e.g. <code>array[array.length-1]</code> for the last item, you can call <code>array.at(-1)</code>. <a href="#Examples">(See the examples below)</a></p>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/typedarray/at/index.html:19]:
```suggestion
<p>This is not to suggest there is anything wrong with using the square bracket notation. For example <code>array[0]</code> would return the first item. However instead of using {{jsxref('TypedArray.prototype.length','array.length')}} for latter items; e.g. <code>array[array.length-1]</code> for the last item, you can call <code>array.at(-1)</code>. <a href="#Examples">(See the examples below)</a></p>
```

File: files/en-us/web/javascript/reference/global_objects/typedarray/at/index.html [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+<div>{{EmbedInteractiveExample("pages/js/typedarray-at.html","shorter")}}</div>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/typedarray/at/index.html:21]:
```suggestion
<div>{{EmbedInteractiveExample("pages/js/typedarray-at.html")}}</div>
```

File: files/en-us/web/javascript/reference/global_objects/typedarray/at/index.html [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+  return arr.at(-1);
+}
+
+const lastItem = returnLast(uint8);
+console.log(lastItem); // Logs: 18
+</pre>
+
+<h3>Comparing methods</h3>
+
+<p>Here we compare different ways to select the penultimate (last but one) item of a {{domxref('TypedArray')}}. Whilst all below methods are valid, it highlights the succinctness and readability of the <code>at()</code> method.</p>
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/global_objects/typedarray/at/index.html:59]:
```suggestion
<p>Here we compare different ways to select the penultimate (last but one) item of a {{jsxref('TypedArray')}}. Whilst all below methods are valid, it highlights the succinctness and readability of the <code>at()</code> method.</p>
```

**Review by Elchi3 — CHANGES_REQUESTED** [review]
Great stuff! Some nits and I think we're good to go.

File: files/en-us/web/javascript/reference/global_objects/array/at/index.html (hunk 9 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/array/at/index.html:18]: Haha - I knew I had used the wrong macro somewhere, but I just couldn't remember where 😆 thank you!

File: files/en-us/web/javascript/reference/global_objects/string/at/index.html (hunk 13 above)
Comment (Rumyra) [reply files/en-us/web/javascript/reference/global_objects/string/at/index.html:66]: 👍 added to my todos

**Review by Elchi3 — APPROVED** [review]
